### PR TITLE
Improved gcode editor

### PR DESCRIPTION
**Pull request by breiler**
This took me insanely long time to get this far... I'm not entirely satisfied yet, but it seems to work.
It addresses the issue #1244 and #1186

It's now possible to save the file and have it reloaded:
Reload on save

I've added some syntax error highlighting. One error is displayed for any missing feed parameter on the first movement-command (right now G1 or G3). Also when connected to GRBL it will check if the commands are supported:
Syntax highlighting

If you have got the time I'd appreciate if you could try it out.

### CODE REVIEW

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/errors/FeedRateMissingErrorParser.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    private int firstMovementLine = 0;
+    private Token<GcodeTokenId> firstMovementToken;
+    private Token<GcodeTokenId> firstFeedRateToken;
+
+    public FeedRateMissingErrorParser(FileObject fileObject) {
+        this.fileObject = fileObject;
+    }
+
+    @Override
+    public void handleToken(Token<GcodeTokenId> token, int line) {
```
Comment (winder) [inline ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/errors/FeedRateMissingErrorParser.java]: Move `firstMovementToken == null` and `firstFeedRateToken == null` to the front of the outer if/else to short circuit the string comparisons for most of the program.

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/SyntaxErrorTask.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import javax.swing.text.Document;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Syntax error scheduler task
+ *
+ * @author Joacim Breiler
+ */
+public class SyntaxErrorTask extends IndexingAwareParserResultTask<GcodeParserResult> {
```
Comment (winder) [inline ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/SyntaxErrorTask.java]: Wow, how did you figure all this out? Lots of moving parts here.

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParserResult.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @author Joacim Breiler
+ */
+public class GcodeParserResult extends ParserResult {
+
+    private final List<GcodeError> errorList;
+
+    public GcodeParserResult(Snapshot snapshot) {
+        super(snapshot);
+        errorList = new ArrayList<>();
```
Comment (winder) [inline ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParserResult.java]: What are the snapshots? If you implemented syntax errors would they be generated as part of the `GcodeParserResult`?

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParser.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * A gcode parser that parses errors from gcode tokens
+ *
+ * @author Joacim Breiler
+ */
+@ServiceProvider(service = GcodeParser.class)
+public class GcodeParser extends Parser {
+
+    private int line;
+    private TokenSequence<GcodeTokenId> tokenSequence;
```
Comment (winder) [inline ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParser.java]: I guess that this is generated by the Lexer? Does Netbeans automatically connect those pieces together and generate `snapshot.getTokenHierarchy()`?

**Comment by winder** [comment]
This feature definitely needs to be more prominent, the "edit" button should definitely be in the file menu (or even the toolbar)

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/VisualizerMultiviewElement.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+/*@MultiViewElement.Registration(
+        displayName = "#LBL_Gcode_Visualiser",
+        iconBase = "com/willwinder/ugs/nbp/editor/edit.png",
+        mimeType = "text/xgcode",
+        persistenceType = TopComponent.PERSISTENCE_ONLY_OPENED,
+        preferredID = "Gcode",
+        position = 1001
+)*/
+@NbBundle.Messages("LBL_Gcode_Visualiser=Visualiser")
+public class VisualizerMultiviewElement extends JPanel implements MultiViewElement {
```
[the remark this answers is not in the document]
Comment (breiler) [reply ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/VisualizerMultiviewElement.java]: Oh, sorry about that. Yes I experimented in integrating the visualizer into the editor. That way the editor would be more prominent. It kind of worked but it's a bit unnecessary and if the user loads a file that is 10Mb or larger the editor eats up all resources... So it really should be optional to load the editor.

I'll remove it!

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/SyntaxErrorTask.java (hunk 2 above)
Comment (breiler) [reply ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/SyntaxErrorTask.java]: As I said. Countless tutorials and documentation and debugging into Netbeans code... I don't even know if there was any documentation that contained all the pieces...

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParser.java (hunk 4 above)
Comment (breiler) [reply ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParser.java]: Yes, exactly!

**Comment by breiler** [comment]
> This feature should be more prominent, the "edit" button should definitely be in the file menu (or even the toolbar)

Yes you are right, I'll revert those changes!

File: ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParserResult.java (hunk 3 above)
Comment (breiler) [reply ugs-platform/ugs-platform-gcode-editor/src/main/java/com/willwinder/ugs/nbp/editor/parser/GcodeParserResult.java]: I actually don't use snapshot, it's used in the super class org.netbeans.modules.csl.spi.ParserResult

**Comment by breiler** [comment]
@winder I know you marked this as approved, but do you feel it's ok to be merged? 
For instance I didn't add any localizations. There could be some more unit test and so on...

**Comment by winder** [comment]
I think this is good, merge it in when you're ready!

**Pull request merged**